qa/tasks/deepsea_deploy.py

```diff
@@ -82,12 +90,21 @@ def __init__(self, ctx, config):
 #       self.log.debug("ctx.config {}".format(ctx.config))
         log.debug("Munged config is {}".format(self.config))
 
+    def _ceph_cluster_status(self):
+        write_file(
```

Comment:
In fact you can reach the same result without copying a file:
```
self.master_remote.run(
    label="ceph cluster status",
    args="""
        set -ex
        ceph pg stat -f json-pretty
        ceph health detail -f json-pretty
        ceph osd tree
        ceph osd pool ls detail -f json-pretty
        ceph -s""")
```
However if we want it to run it as script, I'd add wrapper function to misc,
so it can make sure that it creates unique file on the remote node.

Reply:
@kshtsk I don't really get what you mean by "creates unique file", but please note that `write_file` uses `cat >`, so if the file exists on the remote node it just clobbers the contents.

Reply:
I'm curious how your code will work in practice - I'm using a script because I want each command to be displayed followed by its output, and "set -x" works nicely for this purpose.

Reply:
Hm, ok - it works, but it has two drawbacks:

1. I have to put `sudo` in front of every command. With the script approach, I just create the script on the remote (using `write_file`) and then run it with `sudo bash -c`.
2. The args string looks pretty ugly in the log:

`2018-10-23T08:32:33.329 INFO:teuthology.orchestra.run.target192168000040:Running (ceph cluster status): '\n                set -ex\n                ceph pg stat -f json-pretty\n                ceph health detail -f json-pretty\n                ceph osd tree\n                ceph osd pool ls detail -f json-pretty\n                ceph -s\n                '`

Reply:
> @kshtsk I don't really get what you mean by "creates unique file", but please note that `write_file` uses `cat >`, so if the file exists on the remote node it just clobbers the contents.

I mean that your write_file approach is not thread safe, you will not reproduce in your current use case, but in general if to implement a function like misc.sh_script or misc.sudo_script and there will some parallel usage it can make a trouble.

Reply:
> Hm, ok - it works, but it has two drawbacks:
> 
> 1. I have to put `sudo` in front of every command. With the script approach, I just create the script on the remote (using `write_file`) and then run it with `sudo bash -c`.
You might try:
```
>>> from teuthology.orchestra.remote import Remote as R
>>> R('127.0.0.1').run(args="""sudo -n bash -c "set -ex
... whoami
... uptime
... hostname"
... """)
2018-10-23 11:00:40,607.607 INFO:teuthology.orchestra.run.127:Running: 'true'
2018-10-23 11:00:40,791.791 INFO:teuthology.orchestra.run.127:Running: 'sudo -n bash -c "set -ex\nwhoami\nuptime\nhostname"\n'
2018-10-23 11:00:40,958.958 INFO:teuthology.orchestra.run.127.stderr:+ whoami
2018-10-23 11:00:40,960.960 INFO:teuthology.orchestra.run.127.stdout:root
2018-10-23 11:00:40,960.960 INFO:teuthology.orchestra.run.127.stderr:+ uptime
2018-10-23 11:00:40,964.964 INFO:teuthology.orchestra.run.127.stdout: 11:00:40  up 4 days 20:03,  3 users,  load average: 0.08, 0.06, 0.07
2018-10-23 11:00:40,964.964 INFO:teuthology.orchestra.run.127.stderr:+ hostname
2018-10-23 11:00:40,965.965 INFO:teuthology.orchestra.run.127.stdout:d3
```
I agree, it the better to use remote files in this case, and as I said add utility to not remember or copy paste the right way of running commands.
> 2. The args string looks pretty ugly in the log:
> 
> `2018-10-23T08:32:33.329 INFO:teuthology.orchestra.run.target192168000040:Running (ceph cluster status): '\n set -ex\n ceph pg stat -f json-pretty\n ceph health detail -f json-pretty\n ceph osd tree\n ceph osd pool ls detail -f json-pretty\n ceph -s\n '`

this is actually a bug in orchestra runner, we can improve it...

Reply:
@kshtsk Yeah, I see what you mean. If two threads/processes try to write to the same file at the same time, they will clobber eachother. Case in point: https://tracker.ceph.com/issues/36542

But I can't imagine a sane scenario where two `deepsea_deploy` tasks would be running in parallel.